python/mlx/optimizers/optimizers.py

```diff
@@ -736,3 +736,16 @@ def apply_single(self, gradient: mx.array, parameter: mx.array, state: dict):
         if self.weight_decay != 0:
             parameter += parameter * (-self.weight_decay * learning_rate)
         return parameter - update
+
+
+def clip_grad_norm(grads, max_norm):
+    norm_squared = tree_reduce(
+        lambda acc, g: acc + g.square().sum(), grads, mx.array(0.0)
+    )
+    total_norm = mx.sqrt(norm_squared)
```

Comment:
Slightly nitpicky but I would compute the normalizer here instead of in the clipper function, ie

```python
normalizer = max_norm / (total_norm + 1e-6)
```